File: kanonak-codec/java/src/main/java/org/kanonak/codec/CodecSchema.java
package org.kanonak.codec;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The metadata a generated SDK embeds for its codec — describes the SDK's OWN
 * classes (keyed by their durable URIs) plus the resolved foundation URIs. It
 * does NOT carry a package identity: the classes belong to the SDK's package,
 * but the instances a consumer builds live in the consumer's own (data) package,
 * whose identity is supplied at call time via {@link PackageContext}.
 *
 * <ul>
 *   <li>{@code typePredicate} — resolved {@code kanonak.org/core-rdf@<ver>/type} URI.</li>
 *   <li>{@code labelPredicate} — resolved {@code kanonak.org/core-rdf@<ver>/label} URI.</li>
 *   <li>{@code packageTypeUri} — resolved {@code kanonak.org/core-kanonak@<ver>/Package} URI.</li>
 *   <li>{@code classes} — classes keyed by durable type URI (the node's {@code $type}).</li>
 * </ul>
 */
public record CodecSchema(
    String typePredicate,
    String labelPredicate,
    String packageTypeUri,
    Map<String, CodecClass> classes,
    Map<String, CodecEnum> enums
) {
    /**
     * Source-compatible 4-arg form for callers written before 0.5.0 added
     * {@code enums}. Yields a schema declaring no enumerations.
     */
    public CodecSchema(String typePredicate, String labelPredicate,
                       String packageTypeUri, Map<String, CodecClass> classes) {
        this(typePredicate, labelPredicate, packageTypeUri, classes, Map.of());
    }

    /** A class's canonicalization schema: its durable URI + its (flattened) props. */
    public record CodecClass(String typeUri, Map<String, CodecProp> props) {}

    /**
     * One member of a closed set of named individuals. A record even when
     * {@code label} is its only component, so a later addition (an ordered
     * scale's dominates closure, runtime#16) stays additive rather than a
     * breaking shape change.
     *
     * <p>{@code label} is the ontology's own, read through the same predicate
     * {@link #labelPredicate()} names. Display-lens resolution is a RENDERING
     * concern and deliberately not here.
     */
    public record CodecEnumMember(String label) {}

    /**
     * A closed set of named individuals — the metadata that lets a consumer
     * holding only the schema resolve a member URI it received on the wire,
     * with no generated SDK for the declaring package.
     *
     * <p>Carried, never enforced. A prop ranged at an enumeration in ANOTHER
     * package has no entry here ({@code classes} and {@code enums} both
     * describe the SDK's OWN package), so a codec that validated a {@code $ref}
     * against these members would false-reject valid cross-package data.
     * Absence means "not mine", never "invalid".
     *
     * <p>{@code members} is keyed by durable VERSIONED URI
     * ({@code publisher/package@version/name}), byte-identical to what the wire
     * carries as {@code {"$ref": ...}}. A versionless key would look right and
     * miss every lookup, so the formation is an invariant, not a convention.
     * The local name is NOT duplicated as a key: derive it from the URI with
     * canonical's local-name parser (runtime#17).
     *
     * <p>What makes a set closed is the PRODUCER's decision, taken in the
     * ontology and resolved by the schema layer above; this runtime carries the
     * outcome and names no vocabulary term.
     */
    public record CodecEnum(String typeUri, Map<String, CodecEnumMember> members) {}

    /**
     * Parse the embedded schema JSON (the form the generators emit and the shared
     * vectors carry) into a {@link CodecSchema}. Generated SDKs embed their schema
     * as a JSON string and call this once at load. JDK-only: a minimal embedded
     * JSON parser reads the schema (only strings/objects appear — durable URIs and
     * the class/prop maps), so the generated codec needs no JSON library at runtime.
     * This is the single public deserializer the generated codec binds to.
     */
    @SuppressWarnings("unchecked")
    public static CodecSchema fromJson(String json) {
        Map<String, Object> s = (Map<String, Object>) Json.parse(json);
        Map<String, CodecClass> classes = new LinkedHashMap<>();
        Map<String, Object> rawClasses = (Map<String, Object>) s.get("classes");
        if (rawClasses != null) {
            for (Map.Entry<String, Object> e : rawClasses.entrySet()) {
                Map<String, Object> c = (Map<String, Object>) e.getValue();
                Map<String, CodecProp> props = new LinkedHashMap<>();
                Map<String, Object> rawProps = (Map<String, Object>) c.get("props");
                if (rawProps != null) {
                    for (Map.Entry<String, Object> pe : rawProps.entrySet()) {
                        Map<String, Object> p = (Map<String, Object>) pe.getValue();
                        props.put(pe.getKey(), new CodecProp(
                            (String) p.get("predicate"),
                            (String) p.get("kind"),
                            (String) p.get("datatype"),
                            (String) p.get("range")));
                    }
                }
                classes.put(e.getKey(), new CodecClass((String) c.get("typeUri"), props));
            }
        }
        // Closed sets (0.5.0, runtime#21). Optional: a schema written before
        // this field remains valid and yields an empty map.
        Map<String, CodecEnum> enums = new LinkedHashMap<>();
        Map<String, Object> rawEnums = (Map<String, Object>) s.get("enums");
        if (rawEnums != null) {
            for (Map.Entry<String, Object> e : rawEnums.entrySet()) {
                Map<String, Object> en = (Map<String, Object>) e.getValue();
                Map<String, CodecEnumMember> members = new LinkedHashMap<>();
                Map<String, Object> rawMembers = (Map<String, Object>) en.get("members");
                if (rawMembers != null) {
                    for (Map.Entry<String, Object> me : rawMembers.entrySet()) {
                        Map<String, Object> m = (Map<String, Object>) me.getValue();
                        members.put(me.getKey(), new CodecEnumMember((String) m.get("label")));
                    }
                }
                enums.put(e.getKey(), new CodecEnum((String) en.get("typeUri"), members));
            }
        }
        return new CodecSchema(
            (String) s.get("typePredicate"),
            (String) s.get("labelPredicate"),
            (String) s.get("packageTypeUri"),
            classes,
            enums);
    }

    /**
     * Minimal JDK-only JSON parser for the embedded schema. Strings, objects,
     * arrays, booleans, null, and numbers (as {@link String} tokens) — enough to
     * read the schema the generators emit without a runtime JSON dependency.
     */
    private static final class Json {
        private final String s;
        private int i;

        private Json(String s) {
            this.s = s;
        }

        static Object parse(String s) {
            Json j = new Json(s);
            j.ws();
            Object v = j.value();
            j.ws();
            return v;
        }

        private Object value() {
            char c = s.charAt(i);
            switch (c) {
                case '{': return object();
                case '[': return array();
                case '"': return string();
                case 't': i += 4; return Boolean.TRUE;
                case 'f': i += 5; return Boolean.FALSE;
                case 'n': i += 4; return null;
                default: return number();
            }
        }

        private Map<String, Object> object() {
            Map<String, Object> m = new LinkedHashMap<>();
            i++; ws();
            if (s.charAt(i) == '}') { i++; return m; }
            while (true) {
                ws();
                String key = string();
                ws(); i++; // ':'
                ws();
                m.put(key, value());
                ws();
                char c = s.charAt(i++);
                if (c == '}') return m;
            }
        }

        private List<Object> array() {
            List<Object> a = new ArrayList<>();
            i++; ws();
            if (s.charAt(i) == ']') { i++; return a; }
            while (true) {
                ws();
                a.add(value());
                ws();
                char c = s.charAt(i++);
                if (c == ']') return a;
            }
        }

        private String string() {
            i++; // opening quote
            StringBuilder sb = new StringBuilder();
            while (true) {
                char c = s.charAt(i++);
                if (c == '"') return sb.toString();
                if (c == '\\') {
                    char e = s.charAt(i++);
                    switch (e) {
                        case '"': sb.append('"'); break;
                        case '\\': sb.append('\\'); break;
                        case '/': sb.append('/'); break;
                        case 'b': sb.append('\b'); break;
                        case 'f': sb.append('\f'); break;
                        case 'n': sb.append('\n'); break;
                        case 'r': sb.append('\r'); break;
                        case 't': sb.append('\t'); break;
                        case 'u':
                            sb.append((char) Integer.parseInt(s.substring(i, i + 4), 16));
                            i += 4;
                            break;
                        default: sb.append(e);
                    }
                } else {
                    sb.append(c);
                }
            }
        }

        private String number() {
            int start = i;
            while (i < s.length() && "+-0123456789.eE".indexOf(s.charAt(i)) >= 0) i++;
            return s.substring(start, i);
        }

        private void ws() {
            while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
        }
    }

    /**
     * One property's canonicalization metadata, as embedded by the generator.
     *
     * <ul>
     *   <li>{@code predicate} — the predicate's durable canonical URI (resolved version).</li>
     *   <li>{@code kind} — {@code "datatype"} (typed scalar) vs {@code "object"} (reference/embedded).</li>
     *   <li>{@code datatype} — the datatype's canonical URI (carrier source); present for datatype props.</li>
     *   <li>{@code range} — the range class's canonical URI (0.2.0); optionally present for object
     *       props. Maps an embedded value's fields when the embedded carries no explicit
     *       {@code $type} (range-derived typing: inference only, never materialized as a
     *       statement); may be {@code null}.</li>
     * </ul>
     */
    public record CodecProp(String predicate, String kind, String datatype, String range) {
        /** Convenience constructor for a prop without a range. */
        public CodecProp(String predicate, String kind, String datatype) {
            this(predicate, kind, datatype, null);
        }
    }
}

File: kanonak-codec/csharp/src/Kanonak.Codec/CodecSchema.cs
using System.Collections.Generic;
using System.Text.Json;

namespace Kanonak.Codec
{
    /// <summary>One property's canonicalization metadata, as embedded by the generator.</summary>
    public sealed class CodecProp
    {
        /// <summary>The predicate's durable canonical URI (with resolved version).</summary>
        public string Predicate;

        /// <summary><c>"datatype"</c> vs <c>"object"</c> — decides typed-scalar vs reference/embedded.</summary>
        public string Kind;

        /// <summary>The datatype's canonical URI (carrier source) — present for datatype props.</summary>
        public string Datatype;

        /// <summary>
        /// The range class's canonical URI — present for object props (0.2.0). Maps an
        /// embedded value's fields when the embedded carries no explicit <c>$type</c>
        /// (range-derived typing: inference only, never materialized as a statement).
        /// </summary>
        public string Range;
    }

    /// <summary>A class's canonicalization schema: its durable URI + its (flattened) props.</summary>
    public sealed class CodecClass
    {
        /// <summary>The class's durable canonical URI — the value of the synthesized type triple.</summary>
        public string TypeUri;

        /// <summary>Properties keyed by local name (the wire field name).</summary>
        public Dictionary<string, CodecProp> Props = new Dictionary<string, CodecProp>();
    }

    /// <summary>
    /// The metadata a generated SDK embeds for its codec — the resolved foundation
    /// URIs plus the SDK's own classes, keyed by their durable <c>$type</c> URIs. It
    /// carries no package identity: the instances a consumer builds live in the
    /// consumer's own (data) package, whose identity is supplied at call time via
    /// <see cref="PackageContext"/>.
    /// </summary>
    /// <summary>
    /// One member of a closed set of named individuals. A class even when
    /// <c>Label</c> is its only field, so a later addition (an ordered scale's
    /// dominates closure, runtime#16) stays additive rather than a breaking
    /// shape change.
    /// </summary>
    public sealed class CodecEnumMember
    {
        /// <summary>
        /// The member's own label — the ontology's, read through the same
        /// predicate <c>LabelPredicate</c> names. Display-lens resolution is a
        /// RENDERING concern and deliberately not here.
        /// </summary>
        public string Label;
    }

    /// <summary>
    /// A closed set of named individuals — the metadata that lets a consumer
    /// holding only the schema resolve a member URI it received on the wire,
    /// with no generated SDK for the declaring package.
    /// <para>
    /// Carried, never enforced. A prop ranged at an enumeration in ANOTHER
    /// package has no entry here (Classes and Enums both describe the SDK's OWN
    /// package), so a codec that validated a <c>$ref</c> against these members
    /// would false-reject valid cross-package data. Absence means "not mine",
    /// never "invalid".
    /// </para>
    /// <para>
    /// What makes a set closed is the PRODUCER's decision, taken in the
    /// ontology and resolved by the schema layer above; this runtime carries
    /// the outcome and names no vocabulary term.
    /// </para>
    /// </summary>
    public sealed class CodecEnum
    {
        /// <summary>The enumeration class's durable URI — the same key form Classes uses.</summary>
        public string TypeUri;

        /// <summary>
        /// Members keyed by durable VERSIONED URI
        /// (<c>publisher/package@version/name</c>), byte-identical to what the
        /// wire carries as <c>{"$ref": ...}</c>. A versionless key would look
        /// right and miss every lookup, so the formation is an invariant, not a
        /// convention. The local name is NOT duplicated as a key: derive it from
        /// the URI with canonical's local-name parser (runtime#17).
        /// </summary>
        public Dictionary<string, CodecEnumMember> Members = new Dictionary<string, CodecEnumMember>();
    }

    public sealed class CodecSchema
    {
        /// <summary>Resolved <c>kanonak.org/core-rdf@&lt;ver&gt;/type</c> predicate URI.</summary>
        public string TypePredicate;

        /// <summary>Resolved <c>kanonak.org/core-rdf@&lt;ver&gt;/label</c> predicate URI.</summary>
        public string LabelPredicate;

        /// <summary>Resolved <c>kanonak.org/core-kanonak@&lt;ver&gt;/Package</c> class URI.</summary>
        public string PackageTypeUri;

        /// <summary>Classes keyed by durable type URI (the node's <c>$type</c>).</summary>
        public Dictionary<string, CodecClass> Classes = new Dictionary<string, CodecClass>();

        /// <summary>
        /// Closed sets of named individuals keyed by durable type URI (0.5.0,
        /// runtime#21). Optional, so it is additive. Canonicalization-INERT —
        /// content hashes are unchanged by its presence, which the enums
        /// vectors pin rather than assert. An entry here does NOT require a
        /// twin in <see cref="Classes"/>: a Classes entry is needed only when
        /// instances of that class are emitted as NODES, which an
        /// enumeration's members are not. That absence is load-bearing — it is
        /// what makes an embedded value on an enum-ranged property fail loudly
        /// instead of being mapped as if it were a resource.
        /// </summary>
        public Dictionary<string, CodecEnum> Enums = new Dictionary<string, CodecEnum>();

        /// <summary>
        /// Parse the embedded schema JSON (the form the generators emit and the
        /// shared vectors carry) into a <see cref="CodecSchema"/>. Generated SDKs
        /// embed their schema as a JSON string and call this once at load. Kept
        /// minimal — it is the single public deserializer the generated codec binds to.
        /// </summary>
        public static CodecSchema FromJson(string json)
        {
            using var doc = JsonDocument.Parse(json);
            JsonElement s = doc.RootElement;
            var schema = new CodecSchema
            {
                TypePredicate = s.GetProperty("typePredicate").GetString(),
                LabelPredicate = s.GetProperty("labelPredicate").GetString(),
                PackageTypeUri = s.GetProperty("packageTypeUri").GetString(),
            };
            foreach (var cls in s.GetProperty("classes").EnumerateObject())
            {
                var cc = new CodecClass { TypeUri = cls.Value.GetProperty("typeUri").GetString() };
                foreach (var p in cls.Value.GetProperty("props").EnumerateObject())
                {
                    var prop = new CodecProp
                    {
                        Predicate = p.Value.GetProperty("predicate").GetString(),
                        Kind = p.Value.GetProperty("kind").GetString(),
                    };
                    if (p.Value.TryGetProperty("datatype", out var dt)) prop.Datatype = dt.GetString();
                    if (p.Value.TryGetProperty("range", out var rg)) prop.Range = rg.GetString();
                    cc.Props[p.Name] = prop;
                }
                schema.Classes[cls.Name] = cc;
            }
            // Closed sets (0.5.0, runtime#21). Optional: a schema written
            // before this field remains valid and yields an empty map.
            if (s.TryGetProperty("enums", out var enums))
            {
                foreach (var en in enums.EnumerateObject())
                {
                    var ce = new CodecEnum { TypeUri = en.Value.GetProperty("typeUri").GetString() };
                    if (en.Value.TryGetProperty("members", out var members))
                    {
                        foreach (var m in members.EnumerateObject())
                        {
                            var mem = new CodecEnumMember();
                            if (m.Value.TryGetProperty("label", out var lbl)) mem.Label = lbl.GetString();
                            ce.Members[m.Name] = mem;
                        }
                    }
                    schema.Enums[en.Name] = ce;
                }
            }
            return schema;
        }
    }

    /// <summary>
    /// The identity of the (data) package being content-addressed — the consumer's
    /// package the nodes are assembled into. Used to synthesize the package-wrapper
    /// subject <c>&lt;publisher&gt;/&lt;packageName&gt;@&lt;version&gt;/&lt;packageName&gt;</c>.
    /// </summary>
    public sealed class PackageContext
    {
        public string Publisher;
        public string PackageName;
        public string Version;

        /// <summary>Optional package label (a raw/untyped string statement, as the parser emits).</summary>
        public string Label;
    }
}
